src/main/java/org/opentripplanner/model/plan/Itinerary.java

```diff
@@ -538,14 +540,28 @@ public int getLegIndex(Leg leg) {
   }
 
   /**
-   * The cost of this trip
+   * The fare products of this itinerary.
    */
   public ItineraryFares getFares() {
     return fare;
   }
 
   public void setFare(ItineraryFares fare) {
     this.fare = fare;
+
+    var itineraryInstances = fare
+      .getItineraryProducts()
+      .stream()
+      .map(fp -> {
+        var instanceId = fp.uniqueInstanceId(firstLeg().getStartTime());
+        return new FareProductInstance(instanceId, fp);
+      })
+      .toList();
+
+    this.legs.forEach(l -> {
+        var legInstances = fare.getLegProducts().get(l).stream().toList();
+        l.addFareProducts(ListUtils.combine(itineraryInstances, legInstances));
+      });
```

Comment:
This is mapping code and do not belong inside the Itinerary class, it should be in a mapper class that we could unit test. Also, this made me aware that we have not created "extension points" for the fare model classes included in the itinerary. There is a dependency to the ext packages in the import - witch is not allowed. We should discuss how to do this. My take is to bring the necessary API classes into the OTP api model, they need to be part of the API. We can mark them as "Part of Sandbox".

Reply:
What do you think about adding it to the `FaresFilter` class? This is the kind of glue code that adds the fares to the itinerary and could also do the same for legs.

Yes, lets have a chat in the dev meeting about where the fare classes should live.

Reply:
Suggestion: When we need to change the OTP APIs (internal and external), then these changes need to be in the core. I suggest that we create an annotation(something like `@Sandbox(partOf="Fare")`) and use that to tag all code in the main part of OTP witch play a "Sandbox role". The main principle is to create an extention and put as much logic in the sandbox as possible. This should be injected using dependency injection. API and extensions should be tagged with the `@Sandbox` and is subject to review like normal code. I can clearify/update the Sandbox contract.